<a href="https://colab.research.google.com/github/CervantesMCinthiaK/Metodos-Numericos/blob/main/M%20Steffensen.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Método de Steffensen
Alumna:Cinthia Karina Cervantes Martinez


In [79]:
from math import *
import numpy as np
import sympy as sp #por si se ingresa funciones trigonometricas
import matplotlib . pyplot as plt

In [80]:
#El usuario ingresa la función que desea iterar
funcion=(input("Ingrese la funcion: "))

Ingrese la funcion: (cos(x) + 3)/2


In [81]:
#se define dicha función ingresada por el usuario
def g(x):
  return eval(funcion)

In [82]:

#teniendo en cuenta los intervalos
print("Ingrese el intervalo de la funcion")
a0 = float(input("Ingrese  a: "))
b0 = float(input("Ingresa  b: "))

Ingrese el intervalo de la funcion
Ingrese  a: 1
Ingresa  b: 2


In [83]:
# Ingreso datos de entrada para los diferentes métodos a trabajar
a= float(input("Ingrese el valor de P0(Aproximacion inicial): "))

# guarda valores iniciales
p0 = a

#guarda valores iniciales del error y del número de iteraciones
tol =float(input("Ingrese el valor de la tolerancia: "))
nmax = int(input("Ingrese el número máximo de iteraciones: "))
error = 100
niter = 0

Ingrese el valor de P0(Aproximacion inicial): 1.5
Ingrese el valor de la tolerancia: 0.00001
Ingrese el número máximo de iteraciones: 100


In [84]:
# Método de Steffense
# Cálculo inicial
p1 = g(p0)
p2 = g(p1)
p = p0 - (p1 - p0)**2 / (p2 - 2 * p1 + p0)
#voy imprimiendo la tabla
print("# iter\t\t p0 \t\t g(p) \t\t g(g(p)) \t\t p \t\t  error")
print("{0} \t\t {1:6.6f} \t {2:6.6f} \t {3:6.6f} \t {4:6.6f} \t {5:6.6f}".format(niter, p0, p1, p2, p, error))
# ciclo iterativo
while error > tol and niter < nmax:
    niter += 1

    # Paso 1: Actualizar p0, p1 y p2
    p0 = p  # Actualiza p0 con el nuevo valor
    p1 = g(p0)  # Calcula g(p0) para obtener el nuevo p1
    p2 = g(p1)  # Calcula g(p1) para obtener el nuevo p2

    # Paso 2: Calcular p usando la fórmula de Steffensen
    p = p0 - (p1 - p0)**2 / (p2 - 2 * p1 + p0)


    # Calcular el nuevo error
    error = abs(p - p0)

    # Imprime los resultados de la iteración actual
    print("{0} \t\t {1:6.6f} \t {2:6.6f} \t {3:6.6f} \t {4:6.6f} \t {5:6.6f}".format(niter, p0, p1, p2, p, error))
    # Actualiza p0 para la próxima iteración
    p0 = p

# Verifica si el método funciono o no
if error <= tol:
    print()#dejar un espacio
    print("La raíz de la función dada en el intervalo  [{0:6.4f}, {1:6.4f}] es {2:6.6f}".format(a0,b0,p))
else:
    print(f"El método falló después de {nmax} iteraciones.")

    #se agrego una tabla donde se encuentra p0, p1=g(p0) , p2=g(g(p0)), p y el error.
    #aclarando que la raiz se encuentra en p (donde se evalua la formula de Steffensen)


# iter		 p0 		 g(p) 		 g(g(p)) 		 p 		  error
0 		 1.500000 	 1.535369 	 1.517710 	 1.523591 	 100.000000
1 		 1.523591 	 1.523594 	 1.523592 	 1.523593 	 0.000002

La raíz de la función dada en el intervalo  [1.0000, 2.0000] es 1.523593


Se tiene un ejemplo del código realizado, es decir, usando el método punto fijo se resuelve la función  $$
f(x) = \cos(x) - 2x + 3 = 0
$$
 con una aproximación $ P_0 = 1.5 $ y con un error $10^{-5}$ el cual se reescribió de la siguiente manera:  $$
g(x)=\frac{\cos(x) + 3}{2}
$$
debido a que en esa función se encuentra una raíz dentro del intervalo [1,2]
